_studio/mfx_lib/shared/src/libmfxsw_encode.cpp

```diff
@@ -813,147 +813,160 @@ mfxStatus MFXVideoENCODE_EncodeFrameAsync(mfxSession session, mfxEncodeCtrl *ctr
+                    // fill dependencies
+                    task.pSrc[0] = entryPoints[0].pParam;
+                    task.pDst[0] = ((mfxStatus)MFX_ERR_MORE_DATA_SUBMIT_TASK == mfxRes) ? 0: bs;
+
+    #ifdef MFX_TRACE_ENABLE
+                    task.nParentId = MFX_AUTO_TRACE_GETID();
+                    task.nTaskId = MFX::CreateUniqId() + MFX_TRACE_ID_ENCODE2;
+    #endif
+                    // register input and call the task
+                    MFX_CHECK_STS(session->m_pScheduler->AddTask(task, &syncPoint));
+                }
+
+                // IT SHOULD BE REMOVED
+                if ((mfxStatus)MFX_ERR_MORE_DATA_SUBMIT_TASK == mfxRes)
+                {
+                    mfxRes = MFX_ERR_MORE_DATA;
+                    syncPoint = NULL;
+                }
             }
-            else
+            else if (MFX_WRN_DEVICE_BUSY == mfxRes)
             {
-                MFX_TASK task;
-
-                memset(&task, 0, sizeof(task));
-                task.pOwner = session->m_pENCODE.get();
-                task.entryPoint = entryPoints[0];
-                task.priority = session->m_priority;
-                task.threadingPolicy = session->m_pENCODE->GetThreadingPolicy();
-                // fill dependencies
-                task.pSrc[0] = surface;
-                task.pSrc[1] = ctrl ? ctrl->ExtParam : 0;
-                task.pDst[0] = entryPoints[0].pParam;
-
-#ifdef MFX_TRACE_ENABLE
-                task.nParentId = MFX_AUTO_TRACE_GETID();
-                task.nTaskId = MFX::CreateUniqId() + MFX_TRACE_ID_ENCODE;
-#endif
-                // register input and call the task
-                MFX_CHECK_STS(session->m_pScheduler->AddTask(task, &syncPoint));
-
-                memset(&task, 0, sizeof(task));
-                task.pOwner = session->m_pENCODE.get();
-                task.entryPoint = entryPoints[1];
-                task.priority = session->m_priority;
-                task.threadingPolicy = session->m_pENCODE->GetThreadingPolicy();
-                // fill dependencies
-                task.pSrc[0] = entryPoints[0].pParam;
-                task.pDst[0] = ((mfxStatus)MFX_ERR_MORE_DATA_SUBMIT_TASK == mfxRes) ? 0: bs;
-
-#ifdef MFX_TRACE_ENABLE
-                task.nParentId = MFX_AUTO_TRACE_GETID();
-                task.nTaskId = MFX::CreateUniqId() + MFX_TRACE_ID_ENCODE2;
-#endif
-                // register input and call the task
-                MFX_CHECK_STS(session->m_pScheduler->AddTask(task, &syncPoint));
-            }
+                mfxStatus mfxFindRes = session->m_pScheduler->FindActiveSyncPoint(session->m_pENCODE.get(), &syncPoint);
+                // If we can't find any active SyncPoint's it means that next call to MFXVideoDECODE_DecodeFrameAsync should already accept a task.
+                // Let's repeat the call right here instead of making application do it.
+                if (MFX_ERR_NOT_FOUND == mfxFindRes)
+                     continue;
```

Comment:
It is better to use some counter here(only one repeat-call) to avoid infinite loop **inside** ...FrameAsync() if something works wrong in components logic.

Reply:
Yes, this is a good point. I didn't do it because wan't sure what this max counter should be.

Reply:
Hmm, perhaps the right answer is not "1" but AsyncDepth. I mean a component can hold num tasks not more than AsyncDepth. It means we need to set this the counter  to AsyncDepth. What do you mean?

Reply:
Yeah, AsyncDepth should work.
Though i'm not sure how you can to get it from component on this level.